File: ESP32_nodes/room4/main.py
import network
import time
from machine import Pin, Timer
from umqtt.simple import MQTTClient
import json

# === Node Configuration ===
NODE_ID = 'wc_female_02'  # Matches Room4 (Female WC)
NODE_TYPE = 'female'
ROOM_NAME = 'Room 2'
MQTT_BROKER = '192.168.1.182'  # PC host IP address
MQTT_PORT = 1883

# === GPIO Setup ===
# Status LED (built-in on most ESP32 boards)
status_led = Pin(2, Pin.OUT)
# Relay for controlling actuator (water valve, pump, etc.)
relay = Pin(5, Pin.OUT)
# Optional: Status indicator LED
indicator_led = Pin(18, Pin.OUT)

# Initialize pins
status_led.value(1)  # OFF (inverted on ESP32)
relay.value(0)       # OFF
indicator_led.value(0)  # OFF

# === Global variables ===
mqtt_client = None
relay_timer = None
blink_timer = None
blink_count = 0
wifi_connected = False
mqtt_connected = False

# === WiFi Connection ===
def connect_wifi():
    global wifi_connected
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    
    # Scan for available networks and connect
    networks = [n[0].decode() for n in wlan.scan()]
    
    if "Michelle" in networks:
        wlan.connect('Michelle', '0908800130')
    elif "Vinternal" in networks:
        wlan.connect('Vinternal', 'Veg@s123')
    else:
        print(f"[{NODE_ID}] No known WiFi networks found")
        return False
    
    # Wait for connection
    for i in range(20):
        if wlan.isconnected():
            wifi_connected = True
            ip = wlan.ifconfig()[0]
            print(f"[{NODE_ID}] WiFi connected! IP: {ip}")
            return True
        time.sleep(0.5)
        print(f"[{NODE_ID}] Connecting to WiFi... ({i+1}/20)")
    
    wifi_connected = False
    print(f"[{NODE_ID}] WiFi connection failed")
    return False

# === MQTT Message Handler ===
def on_mqtt_message(topic, message):
    global relay_timer, blink_count
    try:
        # Parse command
        if isinstance(message, bytes):
            msg_str = message.decode().strip()
        else:
            msg_str = str(message).strip()
            
        print(f"[{NODE_ID}] Received: {topic.decode()} -> {msg_str}")
        
        # Try to parse as JSON first
        try:
            cmd_data = json.loads(msg_str)
            action = cmd_data.get('action', msg_str)
        except:
            # Fallback to plain string
            action = msg_str
        
        # Handle flush command
        if action in ['flush', 'on', 'activate']:
            print(f"[{NODE_ID}] FLUSH command received!")
            execute_flush()
            
        # Handle status request
        elif action in ['status', 'ping']:
            publish_status()
            
        # Handle off command
        elif action in ['off', 'stop']:
            stop_relay()
            
    except Exception as e:
        print(f"[{NODE_ID}] Error processing message: {e}")

# === Execute Flush Action ===
def execute_flush():
    global relay_timer, blink_count
    
    # Activate relay (water valve, pump, etc.)
    relay.value(1)
    indicator_led.value(1)
    
    # Start status LED blinking
    blink_count = 25  # Blink for 5 seconds (25 x 200ms)
    start_status_blink()
    
    # Set auto-off timer (5 seconds)
    if relay_timer:
        relay_timer.deinit()
    relay_timer = Timer(-1)
    relay_timer.init(period=5000, mode=Timer.ONE_SHOT, callback=stop_relay)
    
    # Send confirmation back to server
    publish_response("flush", True, "Flush executed successfully")
    
    print(f"[{NODE_ID}] Flush activated for 5 seconds")

# === Status LED Blinking ===
def start_status_blink():
    global blink_count
    if blink_count > 0:
        status_led.value(not status_led.value())
        blink_count -= 1
        # Schedule next blink
        if blink_timer:
            blink_timer.deinit()
        blink_timer = Timer(-1)
        blink_timer.init(period=200, mode=Timer.ONE_SHOT, callback=lambda t: start_status_blink())
    else:
        status_led.value(1)  # OFF (inverted)

# === Stop Relay ===
def stop_relay(timer_obj=None):
    relay.value(0)
    indicator_led.value(0)
    if relay_timer:
        relay_timer.deinit()
        relay_timer = None
    print(f"[{NODE_ID}] Relay stopped")
    
    # Send status update
    publish_response("stop", True, "Relay deactivated")

# === MQTT Connection ===
def connect_mqtt():
    global mqtt_client, mqtt_connected
    try:
        # Create unique client ID
        client_id = f"{NODE_ID}_{time.ticks_ms()}"
        mqtt_client = MQTTClient(client_id, MQTT_BROKER, port=MQTT_PORT)
        mqtt_client.set_callback(on_mqtt_message)
        
        # Connect to broker
        mqtt_client.connect()
        
        # Subscribe to command topic
        command_topic = f"wc/{NODE_ID}/command"
        mqtt_client.subscribe(command_topic)
        
        mqtt_connected = True
        print(f"[{NODE_ID}] MQTT connected and subscribed to {command_topic}")
        
        # Announce online status
        publish_status()
        
        return True
        
    except Exception as e:
        mqtt_connected = False
        print(f"[{NODE_ID}] MQTT connection failed: {e}")
        return False

# === Publish Status ===
def publish_status():
    if mqtt_client and mqtt_connected:
        try:
            status_data = {
                "node_id": NODE_ID,
                "node_type": NODE_TYPE,
                "room_name": ROOM_NAME,
                "status": "online",
                "wifi_connected": wifi_connected,
                "relay_active": relay.value(),
                "timestamp": time.time(),
                "free_memory": str(time.ticks_ms())  # Simple way to show activity
            }
            
            topic = f"wc/{NODE_ID}/status"
            message = json.dumps(status_data)
            mqtt_client.publish(topic, message)
            print(f"[{NODE_ID}] Status published")
            
        except Exception as e:
            print(f"[{NODE_ID}] Failed to publish status: {e}")

# === Publish Response ===
def publish_response(action, success, message=""):
    if mqtt_client and mqtt_connected:
        try:
            response_data = {
                "node_id": NODE_ID,
                "action": action,
                "success": success,
                "message": message,
                "timestamp": time.time()
            }
            
            topic = f"wc/{NODE_ID}/response"
            response_msg = json.dumps(response_data)
            mqtt_client.publish(topic, response_msg)
            
        except Exception as e:
            print(f"[{NODE_ID}] Failed to publish response: {e}")

# === Main Loop ===
def main():
    global mqtt_client
    
    print(f"[{NODE_ID}] Starting {ROOM_NAME} ({NODE_TYPE}) node...")
    
    # Connect to WiFi
    if not connect_wifi():
        print(f"[{NODE_ID}] Failed to connect to WiFi. Retrying in 10 seconds...")
        time.sleep(10)
        return
    
    # Connect to MQTT
    if not connect_mqtt():
        print(f"[{NODE_ID}] Failed to connect to MQTT. Retrying in 10 seconds...")
        time.sleep(10)
        return
    
    print(f"[{NODE_ID}] {ROOM_NAME} node ready!")
    
    # Main loop
    last_status_time = 0
    while True:
        try:
            # Check for MQTT messages
            if mqtt_client:
                mqtt_client.check_msg()
            
            # Send periodic status updates (every 30 seconds)
            current_time = time.time()
            if current_time - last_status_time > 30:
                publish_status()
                last_status_time = current_time
            
            time.sleep_ms(100)
            
        except Exception as e:
            print(f"[{NODE_ID}] Error in main loop: {e}")
            time.sleep(1)
            
            # Try to reconnect if connection lost
            if not wifi_connected:
                connect_wifi()
            if wifi_connected and not mqtt_connected:
                connect_mqtt()

# === Start the node ===
if __name__ == "__main__":
    while True:
        try:
            main()
        except Exception as e:
            print(f"[{NODE_ID}] Critical error: {e}")
            time.sleep(5)

File: ESP32_nodes/room4/src/main.cpp
#include <WiFi.h>
#include <PubSubClient.h>
#include <ArduinoJson.h>

// === Node Configuration ===
constexpr const char* NODE_ID = "wc_female_02";
constexpr const char* NODE_TYPE = "female";
constexpr const char* ROOM_NAME = "Room 2";
constexpr const char* MQTT_BROKER = "192.168.20.109";
constexpr int   MQTT_PORT = 1883;

// === GPIO Setup (adjust pins as needed for your ESP32-C3 board) ===
constexpr int STATUS_LED_PIN = 2;
constexpr int RELAY_PIN      = 5;
constexpr int INDICATOR_PIN  = 8;

constexpr int WIFI_RETRY_MAX = 40;
constexpr int WIFI_RETRY_DELAY = 250;
constexpr int WIFI_SSID_COUNT = 2;
constexpr int STATUS_PUBLISH_INTERVAL = 10000;
constexpr int RELAY_ON_TIME = 5000;

WiFiClient espClient;
PubSubClient client(espClient);

unsigned long lastStatusTime = 0;
unsigned long relayOffTime = 0;
bool relayActive = false;

void setup_pins() {
  pinMode(STATUS_LED_PIN, OUTPUT);
  pinMode(RELAY_PIN, OUTPUT);
  pinMode(INDICATOR_PIN, OUTPUT);
  digitalWrite(STATUS_LED_PIN, HIGH);
  digitalWrite(RELAY_PIN, LOW);
  digitalWrite(INDICATOR_PIN, LOW);
}

void blink_led(unsigned int ms) {
  int cycles = ms / 200;
  for (int i = 0; i < cycles; i++) {
    digitalWrite(STATUS_LED_PIN, LOW);
    delay(100);
    digitalWrite(STATUS_LED_PIN, HIGH);
    delay(100);
  }
  digitalWrite(STATUS_LED_PIN, HIGH);
}

void setup_wifi() {
  delay(10);
  Serial.println("[room4] Starting WiFi connection...");
  WiFi.mode(WIFI_STA);
  WiFi.disconnect(true);
  delay(100);
  pinMode(STATUS_LED_PIN, OUTPUT);
  pinMode(INDICATOR_PIN, OUTPUT);
  digitalWrite(STATUS_LED_PIN, HIGH);
  digitalWrite(INDICATOR_PIN, HIGH); // Indicator LED ON while searching

  const char* ssids[WIFI_SSID_COUNT] = {"Floor 9", "WC"};
  const char* passwords[WIFI_SSID_COUNT] = {"Veg@s123", "abcd123456"};
  bool connected = false;
  while (!connected) {
    for (int net = 0; net < WIFI_SSID_COUNT && !connected; net++) {
      Serial.printf("[room4] Trying SSID: %s\n", ssids[net]);
      WiFi.begin(ssids[net], passwords[net]);
      int retry = 0;
      while (WiFi.status() != WL_CONNECTED && retry < WIFI_RETRY_MAX) {
        delay(WIFI_RETRY_DELAY);
        digitalWrite(STATUS_LED_PIN, !digitalRead(STATUS_LED_PIN));
        digitalWrite(INDICATOR_PIN, !digitalRead(INDICATOR_PIN));
        Serial.print(".");
        retry++;
      }
      if (WiFi.status() == WL_CONNECTED) {
        connected = true;
        break;
      }
      Serial.printf("\n[room4] Failed to connect to %s\n", ssids[net]);
    }
    if (!connected) {
      Serial.println("[room4] No WiFi found, resetting device in 2 seconds...");
      delay(2000);
      ESP.restart();
    }
  }
  digitalWrite(STATUS_LED_PIN, HIGH);
  digitalWrite(INDICATOR_PIN, HIGH);
  Serial.println("\n[room4] WiFi connected!");
  Serial.print("[room4] IP: ");
  Serial.println(WiFi.localIP());
  digitalWrite(INDICATOR_PIN, LOW); // Indicator LED OFF when connected
}

void publish_status(const char* custom_status = nullptr) {
  JsonDocument doc;
  doc["node_id"] = NODE_ID;
  doc["node_type"] = NODE_TYPE;
  doc["room_name"] = ROOM_NAME;
  doc["status"] = custom_status ? custom_status : (WiFi.status() == WL_CONNECTED ? "online" : "offline");
  doc["wifi_connected"] = (WiFi.status() == WL_CONNECTED);
  doc["relay_active"] = relayActive;
  doc["timestamp"] = millis() / 1000;
  char buf[256];
  size_t n = serializeJson(doc, buf);
  String topic = String("wc/") + NODE_ID + "/status";
  client.publish(topic.c_str(), buf, n);
  Serial.printf("[room4] Status published: %s\n", doc["status"].as<const char*>());
}

void publish_response(const char* action, bool success, const char* message) {
  JsonDocument doc;
  doc["node_id"] = NODE_ID;
  doc["action"] = action;
  doc["success"] = success;
  doc["message"] = message;
  doc["timestamp"] = millis() / 1000;
  char buf[256];
  size_t n = serializeJson(doc, buf);
  String topic = String("wc/") + NODE_ID + "/response";
  client.publish(topic.c_str(), buf, n);
  Serial.printf("[room4] Response published: %s (%s)\n", action, success ? "OK" : "FAIL");
}

void stop_relay() {
  digitalWrite(RELAY_PIN, LOW);
  digitalWrite(INDICATOR_PIN, LOW);
  relayActive = false;
  publish_response("stop", true, "Relay deactivated");
  publish_status("idle");
  Serial.println("[room4] Relay stopped, indicator OFF");
}

void execute_flush() {
  digitalWrite(RELAY_PIN, HIGH);
  digitalWrite(INDICATOR_PIN, HIGH);
  relayActive = true;
  relayOffTime = millis() + RELAY_ON_TIME;
  publish_response("flush", true, "Flush executed successfully");
  publish_status("flushing");
  Serial.println("[room4] FLUSH command received! Relay ON, indicator ON, blinking status LED");
  blink_led(RELAY_ON_TIME);
}

void callback(char* topic, byte* payload, unsigned int length) {
  String msg;
  for (unsigned int i = 0; i < length; i++) msg += (char)payload[i];
  String action = msg;
  Serial.print("[room4] MQTT message received: ");
  Serial.println(msg);
  JsonDocument doc;
  DeserializationError err = deserializeJson(doc, msg);
  if (!err) {
    if (doc["action"].is<String>()) action = doc["action"].as<String>();
    Serial.print("[room4] Parsed action: ");
    Serial.println(action);
  }
  action.trim();
  if (action == "flush" || action == "on" || action == "activate") {
    execute_flush();
  } else if (action == "status" || action == "ping") {
    publish_status();
    Serial.println("[room4] STATUS command received!");
  } else if (action == "off" || action == "stop") {
    stop_relay();
    Serial.println("[room4] STOP command received!");
  } else {
    publish_response(action.c_str(), false, "Unknown action");
    Serial.print("[room4] Unknown action: ");
    Serial.println(action);
  }
}

void reconnect() {
  while (!client.connected()) {
    Serial.println("[room4] Attempting MQTT connection...");
    if (client.connect(NODE_ID)) {
      String topic = String("wc/") + NODE_ID + "/command";
      client.subscribe(topic.c_str());
      Serial.print("[room4] Subscribed to topic: ");
      Serial.println(topic);
    } else {
      Serial.println("[room4] MQTT connection failed, retrying...");
      delay(2000);
    }
  }
}

void setup() {
  Serial.begin(115200);
  delay(100);
  Serial.println("[room4] Starting node...");
  setup_pins();
  setup_wifi();
  client.setServer(MQTT_BROKER, MQTT_PORT);
  client.setCallback(callback);
  // Test LED startup
  blink_led(900);
  publish_status("booted");
  Serial.println("[room4] Node ready!");
}

void loop() {
  if (!client.connected()) reconnect();
  client.loop();
  unsigned long now = millis();
  if (relayActive && now > relayOffTime) {
    stop_relay();
  }
  if (now - lastStatusTime > STATUS_PUBLISH_INTERVAL) {
    publish_status();
    lastStatusTime = now;
  }
}
